misc - watersurfaces.Rmd: update to v 1.1

cecileherr opened the pull request:
data exploration for the new version of the watersurfaces layer (1.1)

cecileherr commented:
The html notebook:
[watersurfaces_rendered_html.zip](https://github.com/inbo/n2khab-preprocessing/files/6082348/watersurfaces_rendered_html.zip)

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
````diff
@@ -32,58 +35,182 @@ opts_chunk$set(
+levels(ws_11$DIEPKL)
+```
+
+(please note "≥" are rendered as "=" in the html output)
````
Interesting. Maybe that's also the case when using RStudio viewer to see the table? Does it change when applying `n2khab:::convertdf_enc(from = "latin1")` to the data frame? It's an internal function (documented, see `?convertdf_enc`), also used in `read_soilmap()`, to deal with latin1 characters which cause trouble. If this guess is correct, then that's an idea for `read_watersurfaces()` too.

cecileherr replied:
There seems to be a broader problem.

**Looking at the geometrical layer (hereafter ws_11):**

- In RStudio viewer (default encoding RStudio is UTF-8): 

I can see ws_11$DIEPKL with "≥" and  ws_11$CONNECT with correct latin characters (e.g. WVLC = "ANTMOL0152", CONNECT = "geïsoleerd") as expected.

- In the html output:

I get "=" instead of "≥" but the special latin characters are still correct (e.g. "geïsoleerd"), e.g. for WVLC = "ANTMOL0152":

```
##   WVLC  WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE
## * <fct> <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>  
## 1 ANTM~ <NA>     172032 <NA>  <NA>   <NA>    <NA>     = 2 -~ geïsol~ <NA>   
## # ... with 3 more variables: OPPWVL <dbl>, OMTWVL <dbl>, geom <MULTIPOLYGON
```
But if I try to import the **non geometrical layers**, I get more encoding errors:

```
> read_sf(filepath, layer = "LktCONNECT")

##   Code        Omschr                                                                                                                       
## 1 geÃ¯soleerd niet verbonden met een waterloop                                  
## 2 permanent   het watervlak staat permanent in verbinding met minstens Ã©Ã©n wa~
## 3 periodiek   het watervlak staat tijdelijk (door peilbeheer of droogte) in ver~

>  read_sf(filepath, layer = "LktDIEPKL")

##   Code    Omschr             
## 1 0 - 2 m 0 - < 2 m    
## 2 2 - 4 m â‰¥ 2 - < 4 m
## 3 4 - 6 m â‰¥ 4 - < 6 m
## 4 > 6 m   â‰¥ 6 m
```
`"options = "ENCODING=xxx"` doesn't seem to be avalaible for read_sf with a geopackage. 
(GDAL Message 6: driver GPKG does not support open option ENCODING)

A conversion **from latin1 to UTF-8** is not the way to go ("geÃ¯soleerd", "â\u0089¥ 2 - < 4 m", ...)

A conversion **from UTF-8 to latin1** helps only partly (ok for CONNECT, but DIEPKL becomes NA):
```
> print(n2khab:::convertdf_enc(read_sf(filepath, layer = "LktCONNECT"), 
from = "UTF-8", to = "latin1"))

##   Code       Omschr                                                                                                                        
## 1 geïsoleerd niet verbonden met een waterloop                                   
## 2 permanent  het watervlak staat permanent in verbinding met minstens één water~
## 3 periodiek  het watervlak staat tijdelijk (door peilbeheer of droogte) in verb~

> print(n2khab:::convertdf_enc(read_sf(filepath, layer = "LktDIEPKL"), 
from = "UTF-8", to = "latin1"))

##   Code    Omschr    
## 1 0 - 2 m 0 - < 2 m
## 2 2 - 4 m <NA>     
## 3 4 - 6 m <NA>     
## 4 > 6 m   <NA>
```
And I can reconstruct the html output with a conversion **from UTF-8 to WINDOWS-1252** or **from UTF-8 to UTF-8**:
```
 >print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktCONNECT"), from = "UTF-8", to = "WINDOWS-1252"))

##   Code       Omschr                                                                                                               
## 1 geïsoleerd niet verbonden met een waterloop                                   
## 2 permanent  het watervlak staat permanent in verbinding met minstens één water~
## 3 periodiek  het watervlak staat tijdelijk (door peilbeheer of droogte) in verb~

> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktDIEPKL"), from = "UTF-8", to = "WINDOWS-1252"))

##   Code    Omschr           
## 1 0 - 2 m 0 - < 2 m  
## 2 2 - 4 m = 2 - < 4 m
## 3 4 - 6 m = 4 - < 6 m
## 4 > 6 m   = 6 m
```
But I still do not know how to solve this ;-)

florisvdh replied:
Thanks! I understand your analysis shows that the data are encoded as UTF-8. It is generally used in R (e.g. documentation) & other open-source tools and should not be regarded a problem in se, but it is not built-in on Windows (however in Win 10 it is available as an option AFAIK). From https://developer.r-project.org/Blog/public/2020/05/02/utf-8-support-on-windows/index.html:

> R internally allows strings to be represented in the current native encoding, in UTF-8 and in Latin 1. When interacting with the operating system or external libraries, all these representations have to be converted to native encoding. On Linux and macOS today this is not a problem, because the native encoding is UTF-8, so all Unicode characters are supported. On Windows, the native encoding cannot be UTF-8 nor any other that could represent all Unicode characters.
>
> Windows sometimes replaces characters by similarly looking representable ones (“best-fit”), which often works well but sometimes has surprising results, e.g. alpha character becomes letter a. In other cases, Windows may substitute non-representable characters by question marks or other and R may substitute by \uxxx, \UXXXXXXXX or other escapes.

If you'd like to have the output printed properly in Windows, you could try to do some conversion of specific columns, based on your findings above, directly using `iconv(from = "UTF-8", to = "latin1")` (used by `convertdf_enc()`). It appeared to work for `é`. Best do this OS-specific, as it should not be done in UTF-8 systems --> `if (.Platform$OS.type == "windows") { }`.

Running your code on a Linux system returns the below. I think it is consistent with your output, except that the `≥` UTF-8 symbol is displayed correctly (UTF-8 is the native encoding). I see no solution for that one symbol, except rewriting these factor levels within the function, on Windows: `levels(x) <-`

<details><summary><strong>Code and output</strong></summary>

```r
> ws_diepkl
# A tibble: 4 x 2
  Code    Omschr     
  <chr>   <chr>      
1 0 - 2 m 0 - < 2 m  
2 2 - 4 m ≥ 2 - < 4 m
3 4 - 6 m ≥ 4 - < 6 m
4 > 6 m   ≥ 6 m      
> 
> ws_connect
# A tibble: 3 x 2
  Code       Omschr                                                                                             
  <chr>      <chr>                                                                                              
1 geïsoleerd niet verbonden met een waterloop                                                                   
2 permanent  het watervlak staat permanent in verbinding met minstens één waterloop                             
3 periodiek  het watervlak staat tijdelijk (door peilbeheer of droogte) in verbinding met minstens één waterloop
> 
> # 1) geometrical layer:
> 
> # sf_utf8 <- sf::st_read(filepath_11, stringsAsFactors = TRUE, layer = "Watervlakken", options = "ENCODING=UTF-8")%>% 
> #   filter( WVLC =="ANTMOL0152")
> # options = "ENCODING=UTF-8" cannot be used with a geopackage
> 
> 
> # as imported by default:
> ws_11 %>% 
+     filter( WVLC =="ANTMOL0152")
Simple feature collection with 1 feature and 12 fields
geometry type:  MULTIPOLYGON
dimension:      XY
bbox:           xmin: 207712.2 ymin: 214966.5 xmax: 208091.3 ymax: 215343.7
projected CRS:  Belge 1972 / Belgian Lambert 72
# A tibble: 1 x 13
  WVLC   WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE OPPWVL OMTWVL                       geom
* <fct>  <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>    <dbl>  <dbl>         <MULTIPOLYGON [m]>
1 ANTMO… NA       172032 NA    NA     NA      NA       ≥ 2 -… geïsol… NA      88197.  1532. (((207864.7 215041.2, 207…
> 
> #conversion latin1 - UTF-8
> ws_11_2 <- n2khab:::convertdf_enc(ws_11 , from = "UTF-8", to = "latin1")
> ws_11_2 %>% 
+     filter( WVLC =="ANTMOL0152")
Simple feature collection with 1 feature and 12 fields
geometry type:  MULTIPOLYGON
dimension:      XY
bbox:           xmin: 207712.2 ymin: 214966.5 xmax: 208091.3 ymax: 215343.7
projected CRS:  Belge 1972 / Belgian Lambert 72
# A tibble: 1 x 13
  WVLC   WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE OPPWVL OMTWVL                       geom
* <fct>  <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>    <dbl>  <dbl>         <MULTIPOLYGON [m]>
1 ANTMO… NA       172032 NA    NA     NA      NA       NA     geïsol… NA      88197.  1532. (((207864.7 215041.2, 207…
> 
> ws_11_3 <- n2khab:::convertdf_enc(ws_11 , from = "latin1", to = "UTF-8")
> ws_11_3 %>% 
+     filter( WVLC =="ANTMOL0152")
Simple feature collection with 1 feature and 12 fields
geometry type:  MULTIPOLYGON
dimension:      XY
bbox:           xmin: 207712.2 ymin: 214966.5 xmax: 208091.3 ymax: 215343.7
projected CRS:  Belge 1972 / Belgian Lambert 72
# A tibble: 1 x 13
  WVLC   WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE OPPWVL OMTWVL                       geom
* <fct>  <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>    <dbl>  <dbl>         <MULTIPOLYGON [m]>
1 ANTMO… NA       172032 NA    NA     NA      NA       "â\u0… geÃ¯so… NA      88197.  1532. (((207864.7 215041.2, 207…
> 
> #conversion WINDOWS-1252 - UTF-8
> ws_11_4 <- n2khab:::convertdf_enc(ws_11 , from = "UTF-8", to = "WINDOWS-1252")
> ws_11_4 %>% 
+     filter( WVLC =="ANTMOL0152")
Simple feature collection with 1 feature and 12 fields
geometry type:  MULTIPOLYGON
dimension:      XY
bbox:           xmin: 207712.2 ymin: 214966.5 xmax: 208091.3 ymax: 215343.7
projected CRS:  Belge 1972 / Belgian Lambert 72
# A tibble: 1 x 13
  WVLC   WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE OPPWVL OMTWVL                       geom
* <fct>  <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>    <dbl>  <dbl>         <MULTIPOLYGON [m]>
1 ANTMO… NA       172032 NA    NA     NA      NA       NA     "ge\xe… NA      88197.  1532. (((207864.7 215041.2, 207…
> 
> ws_11_5 <- n2khab:::convertdf_enc(ws_11 , from = "WINDOWS-1252", to = "UTF-8")
> ws_11_5 %>% 
+     filter( WVLC =="ANTMOL0152")
Simple feature collection with 1 feature and 12 fields
geometry type:  MULTIPOLYGON
dimension:      XY
bbox:           xmin: 207712.2 ymin: 214966.5 xmax: 208091.3 ymax: 215343.7
projected CRS:  Belge 1972 / Belgian Lambert 72
# A tibble: 1 x 13
  WVLC   WTRLICHC  HYLAC NAAM  GEBIED KRWTYPE KRWTYPES DIEPKL CONNECT FUNCTIE OPPWVL OMTWVL                       geom
* <fct>  <fct>     <dbl> <fct> <fct>  <fct>   <fct>    <fct>  <fct>   <fct>    <dbl>  <dbl>         <MULTIPOLYGON [m]>
1 ANTMO… NA       172032 NA    NA     NA      NA       â‰¥ 2… geÃ¯so… NA      88197.  1532. (((207864.7 215041.2, 207…
> 
> # 2) non geometrical layers:
> 
> # latin1 to UTF-8: not OK:
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktCONNECT"), from =  "latin1", to ="UTF-8")) 
# A tibble: 3 x 2
  Code        Omschr                                                                                               
  <chr>       <chr>                                                                                                
1 geÃ¯soleerd niet verbonden met een waterloop                                                                     
2 permanent   het watervlak staat permanent in verbinding met minstens Ã©Ã©n waterloop                             
3 periodiek   het watervlak staat tijdelijk (door peilbeheer of droogte) in verbinding met minstens Ã©Ã©n waterloop
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktDIEPKL"), from =  "latin1", to ="UTF-8")) 
# A tibble: 4 x 2
  Code    Omschr              
  <chr>   <chr>               
1 0 - 2 m "0 - < 2 m"         
2 2 - 4 m "â\u0089¥ 2 - < 4 m"
3 4 - 6 m "â\u0089¥ 4 - < 6 m"
4 > 6 m   "â\u0089¥ 6 m"      
> 
> # UTF-8 to latin1: 
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktCONNECT"), from = "UTF-8", to = "latin1"))
# A tibble: 3 x 2
  Code       Omschr                                                                                             
  <chr>      <chr>                                                                                              
1 geïsoleerd niet verbonden met een waterloop                                                                   
2 permanent  het watervlak staat permanent in verbinding met minstens één waterloop                             
3 periodiek  het watervlak staat tijdelijk (door peilbeheer of droogte) in verbinding met minstens één waterloop
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktDIEPKL"), from = "UTF-8", to = "latin1"))
# A tibble: 4 x 2
  Code    Omschr   
  <chr>   <chr>    
1 0 - 2 m 0 - < 2 m
2 2 - 4 m NA       
3 4 - 6 m NA       
4 > 6 m   NA       
> 
> # UTF-8 to WINDOWS-1252: 
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktCONNECT"), from = "UTF-8", to = "WINDOWS-1252"))
# A tibble: 3 x 2
  Code            Omschr                                                                                              
  <chr>           <chr>                                                                                               
1 "ge\xefsoleerd" "niet verbonden met een waterloop"                                                                  
2 "permanent"     "het watervlak staat permanent in verbinding met minstens \xe9\xe9n waterloop"                      
3 "periodiek"     "het watervlak staat tijdelijk (door peilbeheer of droogte) in verbinding met minstens \xe9\xe9n wa…
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktDIEPKL"), from = "UTF-8", to = "WINDOWS-1252"))
# A tibble: 4 x 2
  Code    Omschr   
  <chr>   <chr>    
1 0 - 2 m 0 - < 2 m
2 2 - 4 m NA       
3 4 - 6 m NA       
4 > 6 m   NA       
> 
> # UTF-8 to UTF-8: 
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktCONNECT"), from = "UTF-8", to = "UTF-8"))
# A tibble: 3 x 2
  Code       Omschr                                                                                             
  <chr>      <chr>                                                                                              
1 geïsoleerd niet verbonden met een waterloop                                                                   
2 permanent  het watervlak staat permanent in verbinding met minstens één waterloop                             
3 periodiek  het watervlak staat tijdelijk (door peilbeheer of droogte) in verbinding met minstens één waterloop
> print(n2khab:::convertdf_enc(read_sf(filepath_11, layer = "LktDIEPKL"), from = "UTF-8", to = "UTF-8"))
# A tibble: 4 x 2
  Code    Omschr     
  <chr>   <chr>      
1 0 - 2 m 0 - < 2 m  
2 2 - 4 m ≥ 2 - < 4 m
3 4 - 6 m ≥ 4 - < 6 m
4 > 6 m   ≥ 6 m      
```

</details>

cecileherr replied:
Ok, thanks for checking! SoI think we need to add a specific conversion for the Windows users, indeed. 
I will modify this script to add this conclusion and give the conversion a try and then I will move on to read_watersurfaces itself.

florisvdh replied:
BTW it appears that GeoPackages, as all SQLite databases, only support UTF-8 and UTF-16 text encoding. Further, GDAL uses UTF-8 internally and writes GeoPackages as UTF-8. So my first thought, that it could have been encoded as latin1, can never be the case with GeoPackages (but it can happen with shapefiles).

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
```diff
@@ -32,58 +35,182 @@ opts_chunk$set(
+# ws_10 <- read_sf(file.path(fileman_up("n2khab_data"), "10_raw/watersurfaces/versies/v1.0"),
+#                    stringsAsFactors = TRUE) # n = 86026
+
+ws_11 <- read_sf(file.path(fileman_up("n2khab_data"), "10_raw/watersurfaces/watersurfaces.gpkg"),
+                   stringsAsFactors = TRUE)  # n = 88713
```
```suggestion
ws_11 <- read_sf(file.path(fileman_up("n2khab_data"), "10_raw/watersurfaces/watersurfaces.gpkg"),
                   stringsAsFactors = TRUE, layer = "Watervlakken")  # n = 88713
```

florisvdh reviewed (commented):
Thanks a lot to check this thoroughly, very useful outcomes! :+1: 

Did you also have a look at the other tables in the GeoPackage? From below warning, maybe not:

![afbeelding](https://user-images.githubusercontent.com/19164640/110092787-44f0cf00-7d9a-11eb-9a68-4de81326180e.png)

Included layers:

```r
> library(n2khab)
> filepath <- file.path(fileman_up("n2khab_data"), "10_raw/watersurfaces/watersurfaces.gpkg")
> library(sf)
Linking to GEOS 3.9.0, GDAL 3.1.3, PROJ 7.2.1

> st_layers(filepath)
	- 'VirtualXPath'	[XML Path Language - XPath]
Driver: GPKG 
Available layers:
     layer_name geometry_type features fields
1  Watervlakken Multi Polygon    88713     12
2 LktPEILBEHEER            NA        4      2
3   LktKRWTYPES            NA        2      2
4    LktKRWTYPE            NA       17      2
5    LktFUNCTIE            NA       18      2
6     LktDIEPKL            NA        4      2
7    LktCONNECT            NA        3      2

> read_sf(filepath, layer = "LktPEILBEHEER")
	- 'VirtualXPath'	[XML Path Language - XPath]
# A tibble: 4 x 2
  Code                  Omschr                                                  
  <chr>                 <chr>                                                   
1 instroom geregeld     het instroomdebiet van het watervlak wordt kunstmatig g…
2 afvoer geregeld       het afvoerdebiet van het watervlak wordt kunstmatig ger…
3 aan- en afvoer gereg… zowel de wateraanvoer als -afvoer worden kunstmatig ger…
4 geen peilbeheer       het waterpeil van het watervlak wordt niet kunstmatig  …

```

These tables may be consistent with the factor levels in the data source? Their availability in the data source, if useful, makes it preferrable to use them in `read_watersurfaces()` rather than having to hardcode levels as for version 1.0. Can you have a look?

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
```diff
@@ -1,7 +1,10 @@
+  html_document:
+    toc: yes
+    df_print: paged
   html_notebook:
     number_sections: yes
     code_folding: hide
```
Feel free to drop the `html_notebook` output. The `html_document` is more recommendable indeed. Some of the `html_notebook` settings may be useful to carry over to `html_document`.

cecileherr commented:
> Did you also have a look at the other tables in the GeoPackage? 

Corrected, please see 5638fbf.

> These tables may be consistent with the factor levels in the data source? Their availability in the data source, if useful, makes it preferrable to use them in `read_watersurfaces()` rather than having to hardcode levels as for version 1.0. Can you have a look?

What is the source of the other tables in the GeoPackage? Strange that the categories in the layers are not always consistent with the metadata report and the categories in the geometrical layer (e.g. KRWTYPES). So if we except KRWTYPES, we can use the other tables as extended description for the factors.

Updated html document
[watersurfaces.zip](https://github.com/inbo/n2khab-preprocessing/files/6101494/watersurfaces.zip)

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
```diff
@@ -120,45 +206,58 @@ ws_11$WVLC %>% unique %>% length
 I do not check `NAAM` and `GEBIED` since there are many possible options.
 
-**KRWTYPE**: the codes are correct but there are more codes mentioned in the metadata report
+**KRWTYPE**: the codes are correct but there are more codes mentioned in the metadata report and in the geopackage (layer LktKRWTYPE)
```
It is the case in the current `read_watersurfaces()` as well (17 levels indeed if I saw it right). I believe it comes from either the metadata report (v1.0) or from the reference document where the KRW types were defined.

Given that this is a fixed list of levels, I recommend to use all of them as factor levels. Technically it is not a problem that not all levels are present (yet) in the data. You can do `mutate(., x = factor(x, levels = y))`.

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
```diff
@@ -120,45 +206,58 @@ ws_11$WVLC %>% unique %>% length
 - (`PEILBEHEER`: mentioned in report Leyssen et al. 2020 but not available in the dataset)
 - `HYLAC`: `0` values should be `NA` (empty values, i.e. no code available)
 - `KRWTYPE`, `FUNCTIE`: more levels in the metadata report (Leyssen et al. 2020) than in the dataset 
+- `KRWTYPES`: can not use the corresponding layer LktKRWTYPES because the codes are not the same
```
That is an error in the data source.

florisvdh commented on src/miscellaneous/watersurfaces.Rmd:
````diff
@@ -120,45 +206,58 @@ ws_11$WVLC %>% unique %>% length
 ```
-**DIEPKL**: there is an extra code in the dataset ("> 6 m") and the ordering of 
-the levels could be more logical
+**DIEPKL**: there are extra codes in the dataset ("> 6 m" and "0 - 2 m") and the ordering of 
+the levels could be more logical. 
````
> the ordering of 
the levels could be more logical


That may be a point of attention for `read_watersurfaces()` as well.

florisvdh replied:
Revisiting previous comment (post-hoc): no, we don't do this in the case of raw data sources (at least in the default function behaviour). And such thing hasn't been implemented in `read_watersurfaces()`, so no action needed.

Problems in the data will be returned as-is and should be solved in a future version of the data source. By default we just streamline column names and variable types, we make sure that values referring to `NA` are effectively returned as `NA` in R and we try to avoid some encoding problems.

florisvdh reviewed (approved):
Fine analysis @cecileherr!

> What is the source of the other tables in the GeoPackage?

I did not investigate.

>  if we except KRWTYPES, we can use the other tables as extended description for the factors.

:+1: Also, you can consider the fact that `'veedrenk'` cannot be translated (hence: translated as `NA` with a `left_join()`) as an issue of the data source, which the function does not has to solve. The function should return the data source in a reasonable way, but does not need to solve each data problem.

florisvdh commented:
@cecileherr I have no remaining comments on this PR, but I'll leave it to for you to decide on its merging. So you can further add to the branch if you still need it.

cecileherr commented:
Last update: deal with encoding errors in Windows:
e9b79e6
[watersurfaces.zip](https://github.com/inbo/n2khab-preprocessing/files/6102997/watersurfaces.zip)

florisvdh commented:
Thanks for the update!

> we might be better using an alternative such as ">="

OK - for `read_watersurfaces()` please apply such change on all platforms, in order not to get differences between platforms (cf. consistent derived results).
